### DISCUSSION TITLE
Adding a "return if" statement

### DISCUSSION THREAD

**Discussion by OFark** [dotnet/csharplang#6345] [Language Ideas]
So it occurs to me that this could be simpler:
```cs
public Thing? GetSomeThing(int thingId)
{
    var result = respository.GetThing(thingId);
    if(result is null || result.FailsValidation())
    {
        return null;
    }

   do Stuff with result;
}
```

For code simplicity, readability and to reduce arguments I propose the following:

```cs
public Thing? GetSomeThing(int thingId)
{
    var result = respository.GetThing(thingId);
    return if result is null || result.FailsValidation() : null;

   do Stuff with result;
}
```
This return:
- would only return if the condition is met.
- would return null in this case.
- is similar to the `ArgumentNullException.ThrowIfNull(thingId)` one liner;

Like an `else if` statement you can have multiple else if's with a final else. You can have multiple `return if` statements before a final return.

it could even go as far as:
 ```cs
public Thing? GetSomeThing(int thingId)
{
    return if respository.GetThing(thingId) is not Result result : null;

   do Stuff with result;
}
```

**Comment by HaloFour** [comment]
See: #176

**Reply by OFark** [reply]
Good find. I did spend a while searching but I was searching for "return if" not "?? return". Though I would still argue that having the command statement at the beginning of the line makes it clearer what's happening.

**Comment by HaloFour** [comment]
Also, see #2195

**Comment by Muchaszewski** [comment]
I don't know why so many people are so keen on just returning the value in a single line, even considering that for a single line `if` you do not add braces. Consider the following code, which we want to check if `IsValidLong` is valid before continuing.

```
IActionResult? IsValidLong(){
   Thread.Sleep(1000);
   return new BadRequestObject();
}
```

**Example 1:**
```
public IActionResult SomeApi()
{
   if(IsValidLong() != null) return IsValidLong();
}
```

**Example 2:**
```
public IActionResult SomeApi()
{
   var isValid = IsValidLong();
   if(isValid != null) return isValid;
}
```

Or in the proper recommended Microsoft code style for if statements this would look like this:

**Example 1:**
```
public IActionResult SomeApi()
{
   if(IsValidLong() != null) 
   {
       return IsValidLong();
   }
}
```

**Example 2:**
```
public IActionResult SomeApi()
{
   var isValid = IsValidLong();
   if(isValid != null) 
   {
       return isValid;
   }
}
```

Which one should we choose? All of them are bad - reading-wise, both create some kind of waste. We don't need this isValid value stored if it's null, we don't need to pollute our variables.

That's why I think their proper use cases to generate valid C# code, that would return value from an if statement

**Example:**
```
public IActionResult SomeApi()
{
   // or any other proposed variant that would allow doing this is a real one-line style.
   return if (IsValidLong() != null);
}
```

**Reply by HaloFour** [reply]
> I don't know why so many people are so keen on just returning the value in a single line, even considering that for a single line `if` you do not add braces.

Because the braces and whitespace are optional on purpose.  Those style guidelines are just that, guidelines.  They're not intended to be enforced in all cases, especially where it doesn't make sense, like in this case.  If that were the case it would be a compiler error to exclude the braces, but it's not.

> Which one should we choose?

The form that has been _idiomatic_ C# (and Java and C++ and C) since inception.

**Comment by CyrusNajmabadi** [comment]
> Or in the proper recommended Microsoft code style for if statements this would look like this:

This is not the ms code style for if-statements.  We allow the braces to be left off, and we allow it to be a single line. You can see examples of both in roslyn itself, as well as the runtime.

**Comment by daharper** [comment]


**Reply by CyrusNajmabadi** [reply]
@daharper you can already write it in a single line:

`if (x == -1) return;`

There's no need or recommendation to write such a simple construct as 4 lines as you have.

**Reply by daharper** [reply]
Hi @CyrusNajmabadi apologies I didn't see your reply, I was editing my post (a few times) and realised, it's not a problem of the language (as your reply correctly points out) but of convention - many teams mandate you must use braces on all conditions, and enforce that practice in code reviews and tools like Resharper or other linting tools. That's why I deleted my post. However, thank you kindly for the reply.

**Reply by CyrusNajmabadi** [reply]
Yup yup. For context, we generally don't design the language to workaround explicit external conventions.  Especially because the next thing we'll know, that same group will disallow the new feature as well.
